### PR TITLE
Introducing new listener for claim value encryption

### PR DESCRIPTION
**Pull request by lashinijay**
### Proposed changes in this pull request

> Please refer to the $subject

### CODE REVIEW

File: features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml [hunk 1]
```diff
@@ -702,6 +702,7 @@
 				in your user store must be configured for this. -->
 				<AttributeID>totpSecretkey</AttributeID>
 				<Description>Claim to store the secret key</Description>
+				<EnableEncryption>true</EnableEncryption>
```
[the remark this answers is not in the document]
Comment (lashinijay) [reply features/claim-mgt/org.wso2.carbon.claim.mgt.server.feature/resources/conf/claim-config.xml:705]: working on migration script

File: components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/IdentityClaimValueEncryptionListener.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+     * @return true if encryption is enabled for the claim.
+     * @throws UserStoreException
+     * */
+    private boolean checkEnableEncryption(String claimURI, UserStoreManager userStoreManager)
+            throws UserStoreException {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(userStoreManager.getTenantId());
+        Map<String, String> claimProperties = getClaimProperties(tenantDomain, claimURI);
+        Boolean enableEncryption = false;
+        if (!claimProperties.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (lashinijay) [reply components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/IdentityClaimValueEncryptionListener.java]: Yes it is required when the user is removing the configured totp, the secret key is set to an empty string, however, that empty string is getting encrypted generating a value for that, when the user is reconfiguring totp it's checked whether the secret claim is empty before regenerating a secret key. and if there is a value present for the claim, the secret key regenarating is failing.

**Review by pulasthi7 — APPROVED** [review]

File: components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/internal/IdentityMgtServiceComponent.java [hunk 3]
```diff
@@ -378,4 +383,34 @@ protected synchronized void unsetUserOperationEventListenerService(
...
+        if (log.isDebugEnabled()) {
+            log.debug("claimManagementService set in IdentityMgtServiceComponent bundle");
+        }
+        IdentityMgtServiceDataHolder.setClaimManagementService(claimManagementService);
+    }
+
+    /**
+     * Unset claim metadata management service implementation.
+     */
+    protected void unsetClaimMetadataManagementService(ClaimMetadataManagementService claimManagementService) {
```
Comment (indeewari) [inline components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/internal/IdentityMgtServiceComponent.java:409]: No usage "claimManagementService" ?

File: components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/IdentityClaimValueEncryptionListener.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            LOG.error("Error while retrieving tenant ID", e);
+            return false;
+        } catch (CryptoException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean doPreAddUserWithID(String userID, Object credential, String[] roleList, Map<String, String> claims,
```
Comment (indeewari) [inline components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/IdentityClaimValueEncryptionListener.java:80]: "try {
            if (!isEnable() || userStoreManager == null) {
                return true;
            }
            updateClaimValues(claims, userStoreManager);
            return true;
        } catch (org.wso2.carbon.user.api.UserStoreException e) {
            LOG.error("Error while retrieving tenant ID", e);
            return false;
        } catch (CryptoException e) {
            LOG.error("Error occurred while encrypting claim value", e);
            return false;
        }"

If this block keeps repeating in several methods, we could have call a single method with logic via other methods

**Review by indeewari — APPROVED** [review]

**Pull request merged**